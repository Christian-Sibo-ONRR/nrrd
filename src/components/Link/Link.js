import React from 'react'
import PropTypes from 'prop-types'
import { Link as GatsbyLink, withPrefix } from 'gatsby'
import { makeStyles, useTheme, Box } from '@material-ui/core'
import { IconDownloadXlsImg, IconDownloadCsvImg, IconDownloadDataImg, IconDownloadBaseImg } from '../images'

const useStyles = makeStyles(theme => ({
  link: {
    color: theme.palette.text.secondary,
    textDecoration: 'underline',
    '&:hover': {
      textDecoration: 'none',
    }
  },
  headerLink: {
    color: theme.typography.body2.color,
    textDecoration: 'none',
    marginLeft: theme.spacing(2),
    '&:hover': {
      textDecoration: 'underline',
    }
  },
  headerLinkBold: {
    fontWeight: theme.typography.fontWeightBold
  },
})
)

const IconLink = ({ icon, children, pl = 4, ...props }) => (
  <Box pl={0} mt={2} mb={2}>
    <BaseLink {...props} disableRouting>
      <Box mr={1} display='inline-block'>{icon}</Box>
      <span>{children}</span>
    </BaseLink>
  </Box>
)

const BaseLink = ({ href, disableRouting, className = '', children, ...props }) => {
  const theme = useTheme()
  const styles = useStyles(theme)

  const currentPathname = typeof window !== 'undefined' ? window.location.pathname : ''

  let url = href

  const isRelative = (url.charAt(0) !== '#' && !url.includes('http') && !url.includes('mailto'))
  url = isRelative ? withPrefix(url) : url

  const classes = (props.linkType === LinkTypeComponents.Header)
    ? `${ styles.headerLink } ${ className } ${ (currentPathname === href) && styles.headerLinkBold }`
    : `${ styles.link } ${ className }`

  return (
    <React.Fragment>
      {(!disableRouting && isRelative)
        ? <GatsbyLink to={url} className={classes} {...props} >
          {children}
        </GatsbyLink>
        : <a href={url} className={classes} {...props}>
          {children}
        </a>
      }
    </React.Fragment>
  )
}

const LinkTypeComponents = {
  default: props => <BaseLink {...props} />,
  Header: props => <BaseLink {...props} linkType={LinkTypeComponents.Header} />,
  DownloadXls: props => <IconLink icon={<IconDownloadXlsImg />} {...props} />,
  DownloadCsv: props => <IconLink icon={<IconDownloadCsvImg />} {...props} />,
  DownloadData: props => <IconLink icon={<IconDownloadDataImg />} {...props} />,
  DownloadBase: props => <IconLink icon={<IconDownloadBaseImg />} pl={0} {...props} />
}

const regexXlsx = RegExp('.xlsx$')
const regexCsv = RegExp('.csv$')
const regexDownloadData = RegExp('^/downloads/[a-zA-Z0-9]+')

const getLinkComponent = ({ linkType, ...props }) => {
  if (linkType) {
    return LinkTypeComponents[linkType](props)
  }
  if (regexXlsx.test(props.href)) {
    return LinkTypeComponents.DownloadXls(props)
  }
  if (regexCsv.test(props.href)) {
    return LinkTypeComponents.DownloadCsv(props)
  }

  if (props.linkType === 'DownloadBase') {
    return LinkTypeComponents.DownloadBase(props)
  }
  else if (regexDownloadData.test(props.href)) {
    return LinkTypeComponents.DownloadData(props)
  }

  return LinkTypeComponents.default(props)
}

/**
 * This Link component includes logic to determine if we need to use the Gatsby Link. By using Gatsby Link we leverage the
 * internal routing table and the ability to prefetch relative links.
 * By default all relative urls will use Gatsby, however files that need to be downloaded or anchor links do not use Gatsby Link.
 *
 * We also determine a link type by default to assign the appropriate icon.
 */
const Link = props => getLinkComponent(props)

Link.propTypes = {
  /** The url string for the link. */
  href: PropTypes.string.isRequired,
  /**
   * Will prepend the appropriate icon and the use a default routing option for that type. For example download file types do not use routing by default.
   *
   * By default we determine the appropriate link type but you can specify a type if you want to override it.
   */
  linkType: PropTypes.oneOf(['DownloadXls', 'DownloadCsv', 'DownloadData', 'DownloadBase', 'Header', 'default']),
  /**
   * Used to flag a relative link that we may not want to use Gatsby Routing for. An example is download files.
   *
   * By default we determine the appropriate routing but you can use this flag if you want to override it.
   */
  disableRouting: PropTypes.bool
}

Link.defaultProps = {
  disableRouting: false,
}

export default Link
