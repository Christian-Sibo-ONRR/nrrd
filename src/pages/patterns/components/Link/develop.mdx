---
title: Link Component
layout: pattern-library
---

# Properties:

**Path to file:** Link/Link.js

This Link component includes logic to determine if we need to use the Gatsby Link. By using Gatsby Link we leverage the internal routing table and the ability to prefetch relative links. By default all relative urls will use Gatsby, however files that need to be downloaded or anchor links do not use Gatsby Link.
We also determine a link type by default to assign the appropriate icon.

```live
<Link>Link</Link>
```