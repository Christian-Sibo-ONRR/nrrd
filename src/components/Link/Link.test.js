/* eslint-disable no-undef */
import React from 'react'
import Link from './Link'
import { create } from 'react-test-renderer'

describe('Link component:', () => {
  test('External Link Match snapshot', () => expect(create(<Link href="https://revenuedata.doi.gov/">NRRD</Link>).toJSON()).toMatchSnapshot())
  test('DownloadXls Link Match snapshot', () => expect(create(<Link href="./excelfile.xlsx">NRRD</Link>).toJSON()).toMatchSnapshot())
  test('HowWorks Link Match snapshot', () => expect(create(<Link linkType='HowWorks' href="./how-works/native/">NRRD</Link>).toJSON()).toMatchSnapshot())
})
