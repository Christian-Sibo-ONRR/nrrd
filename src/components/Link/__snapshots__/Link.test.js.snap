// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link component: DownloadXls Link Match snapshot 1`] = `
<div
  className="MuiBox-root MuiBox-root-5"
>
  <a
    className="makeStyles-link-1 "
    href="./excelfile.xlsx"
  >
    <div
      className="MuiBox-root MuiBox-root-6"
    >
      <test-file-stub
        style={
          Object {
            "fill": "currentColor",
            "verticalAlign": "middle",
          }
        }
      />
    </div>
    <span>
      NRRD
    </span>
  </a>
</div>
`;

exports[`Link component: External Link Match snapshot 1`] = `
<a
  className="makeStyles-link-1 "
  href="https://revenuedata.doi.gov/"
>
  NRRD
</a>
`;

exports[`Link component: HowWorks Link Match snapshot 1`] = `
<div
  className="MuiBox-root MuiBox-root-7"
>
  <a
    className="makeStyles-link-1 "
    href="./how-works/native/"
  >
    <div
      className="MuiBox-root MuiBox-root-8"
    >
      <test-file-stub
        style={
          Object {
            "fill": "currentColor",
            "verticalAlign": "middle",
          }
        }
      />
    </div>
    <span>
      NRRD
    </span>
  </a>
</div>
`;
